src/metabase/models/user_key_value.clj

```diff
@@ -0,0 +1,121 @@
+   [:value [:maybe :string]]]
+  ```
+
+  Or you could go even further, and define a `:multi` spec that has a spec for keys in the case of particular values:
+
+  ```
+  [:multi {:dispatch :key}
+   [\"string-key\" [:map [:value [:maybe :string]]]]
+   [\"number-key\" [:map [:value [:maybe :int]]]]]
+  ```
+
+  Note that: `value` must always be `:maybe` - a null value means to delete the KVP.
+  "
+  (:require
+   [malli.core :as mc]
+   [malli.transform :as mtx]
+   [metabase.models.user-key-value.types :as types]
+   [metabase.util.malli :as mu]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(methodical/defmethod t2/table-name :model/UserKeyValue [_model] :user_key_value)
+
+(doto :model/UserKeyValue
+  (derive :metabase/model)
+  (derive :hook/timestamped?))
+
+(mu/defn put!
+  "Upserts a KV-pair"
+  [user-id :- :int
+   kvp :- ::types/user-key-value]
+  (let [{:keys [namespace key value expires-at]}
+        (mc/encode ::types/user-key-value
+                   kvp
+                   (mtx/transformer
+                    (mtx/default-value-transformer)
+                    {:name :database}))]
+    (t2/with-transaction [_]
+      (if (t2/select-one :model/UserKeyValue :user_id user-id :namespace namespace :key key)
+        (t2/update! :model/UserKeyValue :user_id user-id :namespace namespace :key key {:value value
+                                                                                        :expires_at expires-at})
+        (try
+          (t2/insert! :model/UserKeyValue {:user_id user-id
+                                           :namespace namespace
+                                           :key key
+                                           :value value
+                                           :expires_at expires-at})
+          ;; in case we caught a duplicate key exception (a row was inserted between our read and write), try updating
+          (catch Exception _
+            (t2/update! :model/UserKeyValue :user_id user-id :namespace namespace :key key {:value value
+                                                                                            :expires_at expires-at})))))
+    value))
+
+(mu/defn delete!
+  "Deletes a KV-pair"
+  [user-id :- :int
+   namespace :- :string
+   k :- :string]
+  (t2/delete! :model/UserKeyValue :namespace namespace :user_id user-id :key k))
+
+(mu/defn retrieve
+  "Retrieves a KV-pair"
+  [user-id :- :int
+   namespace :- :string
+   k :- :string]
+  (when-let [ukv
+             (t2/select-one :model/UserKeyValue
+                            {:where
+                             [:and
+                              [:= :user_id user-id]
+                              [:= :namespace namespace]
+                              [:= :key k]
+                              [:or
+                               [:>= :expires_at :%now]
+                               [:= :expires_at nil]]]})]
```

Comment:
I might be missing it, but are we deleting expired kv pairs anywhere, or just ignoring them during retrieval?

Reply:
Discussed offline — we'll leave out expired KV pair deletion for now, since the initial use cases shouldn't be generating keys dynamically. But this can come in another PR if the use cases expand

Reply:
👍